odoo/addons/base/tests/test_ir_mail_server_smtpd.py

```diff
@@ -0,0 +1,443 @@
+
+        # create a few certificates for ssl/tls+starttls and mock the
+        # various smtplib classes into trusting the created root CA.
+        _logger.info("Using openssl to generate fake certificates, show "
+                     "subprocess output with: --log-handler %s:DEBUG", __name__)
+        tempdir = tempfile.TemporaryDirectory(prefix='odoo-test-smtpd-')
+        cls.addClassCleanup(tempdir.cleanup)
+        cls.ssl_ca, cls.ssl_client, cls.ssl_server, cls.ssl_self_signed = \
+            _generate_certificates(Path(tempdir.name))
+
+        class TEST_SMTP(smtplib.SMTP):
+            def starttls(self, *, context):
+                if context is None:
+                    context = ssl._create_stdlib_context()  # what SMTP_SSL does
+                    # context = ssl.create_default_context()  # what it should do
+                context.load_verify_locations(cafile=str(cls.ssl_ca.cert))
+                super().starttls(context=context)
+        patcher = patch('smtplib.SMTP', TEST_SMTP)
+        patcher.start()
+        cls.addClassCleanup(patcher.stop)
+
+        class TEST_SMTP_SSL(smtplib.SMTP_SSL):
+            def _get_socket(self, *args, **kwargs):
+                # self.context = ssl.create_default_context()  # what it should do
+                self.context.load_verify_locations(cafile=str(cls.ssl_ca.cert))
+                return super()._get_socket(*args, **kwargs)
+        patcher = patch('smtplib.SMTP_SSL', TEST_SMTP_SSL)
+        patcher.start()
+        cls.addClassCleanup(patcher.stop)
+
+        # reactivate sending emails during this test suite, make sure
+        # NOT TO send emails using another ir.mail_server than the one
+        # created in setUp!
+        patcher = patch.object(cls.registry['ir.mail_server'], '_is_test_mode')
+        mock = patcher.start()
+        mock.return_value = False
+        cls.addClassCleanup(patcher.stop)
+
+        # fix runbot, docker uses a single ipv4 stack but it gives ::1
+        # when resolving "localhost" (so stupid), use the following to
+        # force aiosmtpd/odoo to bind/connect to a fixed ipv4 OR ipv6
+        # address.
+        family, _, cls.port = _find_free_local_address()
+        cls.localhost = getaddrinfo('localhost', cls.port, family)
+        cls.startClassPatcher(patch('socket.getaddrinfo', cls.getaddrinfo))
+
+    @classmethod
+    def getaddrinfo(cls, host, port, *args, **kwargs):
+        """
+        Resolve both "localhost" and "notlocalhost" on the ip address
+        bound by aiosmtpd inside `start_smtpd`.
+        """
+        if host in ('localhost', 'notlocalhost') and port == cls.port:
+            return cls.localhost
+        return getaddrinfo(host, port, family=0, type=0, proto=0, flags=0)
+
+    def setUp(self):
+        super().setUp()
+        # Add an outgoing mail server for the server we start with ``start_smtpd``
+        self.mail_server = self.env['ir.mail_server'].create({
```

Comment:
OOC, why not doing it in setUpClass ?

Reply:
Alternatively, we could create one in each test as anyway configuration varies in most of them (a write lost in the middle of the test does not make it super clear what we are testing, compared to a plain create with test values). As you prefer, does not change much in the end.

Reply:
removed the setUp and moved `ir_mail_server.create({...})` to every test